### PR TITLE
GlyphLayout, improved glyph run bounds and whitespace wrapping.

### PR DESCRIPTION
**Pull request by NathanSweet**
**Improves glyph run bounds**
The bounds can still overlap due to kerning, but they are no longer off by an entire character. Instead of using a low x and higher xadvance to position the glyph in the right place, we now increase the x position to the start of the glyph and decrease the xadvance. The test was changed to help differentiate bounds from different runs.
Before:

After:

**Fixes bounds when whitespaces wraps**
This text has "T e". When the wrapping occurs so the space wraps, the bounds of the first wrapped glyph can be wrong. This was because the `if (wrapIndex == 0)` code path is taken instead of `wrap()` and it didn't clear `lastGlyph`.
Before:

After:

**Fixes wrapping whitespace before a newline**
This text is "AA BB \nEE". When the wrapping occurs so the space after "BB" wraps, a blank line appears. This was because we moved down a line for the wrapping, then `newLine` was true and moved down an extra line.
Before:

After:

Would be great to get eyes on this before merging. Maybe @TCreutzenberg and @tommyettinger are interested?

### CODE REVIEW

**Comment by NathanSweet** [comment]
All the commits in the PR show I clearly don't know what I'm doing (guess I should have squashed before the PR), but the 2 files changed is right. We'll just be sure to squash before merging!

**Comment by tommyettinger** [comment]
It looks like this definitely helps with word wrapping when there's a line break. What if there are two or more line breaks where it wraps? I'd think the initial whitespace would be ignored and later line breaks should be respected, right? The rest of this seems to clean up bounds all over, which may help so we don't need a patchwork of fixes with epsilon and font metrics. I like this whole PR, I'll probably approve in a review.

I always have a ton of commits backlogged in anything I do too. Not sure how to get Git to avoid that...

**Review by tommyettinger — APPROVED** [review]
This looks good, but I can't quite follow the logic when reading it as a git diff. The examples on the initial PR post show really solid improvement, so I'm approving. I do do encourage testing with multiple newlines and/or mixed newlines and lines full of all whitespace for wrapping; there are always some corner cases with text rendering that seem to pop up.

**Comment by TCreutzenberg** [comment]
@NathanSweet Just tried it with my fonts and it works great. Yes, the kerning-overlap can't be avoided.
Code looks good.

If I may make one suggestion, I'd change the text in BitmapFontTest to:

`text = "AA BB    [RED]EE[] T [GREEN]e[] V[YELLOW]a bb[] ([CYAN]5[]FFF)";`

This includes a color-run with a space in it and the color-run which made problems for me 2 days ago. So basically it would cover more edgecases and past problems.

**Comment by TCreutzenberg** [comment]
@NathanSweet I just created a new PR (https://github.com/libgdx/libgdx/pull/6582https://github.com/libgdx/libgdx/pull/6582) before looking at this one, it may lead to merge conflicts with this one, so you may want to include the comments into yours, then I can just close mine?

**Comment by NathanSweet** [comment]
> What if there are two or more line breaks where it wraps?
She works properly in that case. Eg for two newlines you get one newline taken care of by the wrapping, then the other one gives a blank line as normal.

> I can't quite follow the logic when reading it as a git diff.
Indeed, it's hard to follow in an IDE with a debugger!

> Oh, and thank you btw!
Sure, no problem! Glad to have more eyes on this code.

I wish we could have an elegant way of fixing this one:
https://github.com/libgdx/libgdx/issues/4672
